### PR TITLE
[dashboard] Add `RAY_CLUSTER_ACTIVITY_HOOK` to `/api/component_activities`

### PR DESCRIPTION
**Pull request by nikitavemuri**
## Why are these changes needed?

- Add external hook to `/api/component_activities` endpoint in dashboard snapshot router
- Change `is_active` field of `RayActivityResponse` to take an enum `RayActivityStatus` instead of `bool`. This is a backward incompatible change, but should be ok because https://github.com/ray-project/ray/pull/25996 wasn't included in any branch cuts. `RayActivityResponse` now supports informing when there was an error getting the activity observation and the reason.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/snapshot/tests/test_snapshot.py [hunk 1]
```diff
@@ -1,29 +1,109 @@
...
+    class MockClassMethodRouteTable:
+        @classmethod
+        def get(cls, path, **kwargs):
+            return lambda x: x
+
+    patch(
+        "ray.dashboard.optional_utils.ClassMethodRouteTable", MockClassMethodRouteTable
+    ).start()
+    mock_load_class = Mock(return_value=Mock(return_value=cluster_activity_hook_output))
+    os.environ[ray_constants.RAY_CLUSTER_ACTIVITY_HOOK] = "mock_module"
```
Comment (rkooo567) [inline dashboard/modules/snapshot/tests/test_snapshot.py:67]: can you test with a function that dynamically changes the output instead of a static output?

Comment (nikitavemuri) [reply dashboard/modules/snapshot/tests/test_snapshot.py]: yes, updated

File: dashboard/modules/snapshot/snapshot_head.py [hunk 2]
```diff
@@ -108,16 +114,40 @@ async def snapshot(self, req):
...
             timeout = int(timeout)
         else:
             timeout = 5
 
+        # Get activity information for driver
         driver_activity_info = await self._get_job_activity_info(timeout=timeout)
 
+        external_ray_cluster_activity_output = {}
+        if ray_constants.RAY_CLUSTER_ACTIVITY_HOOK in os.environ:
+            external_ray_cluster_activity_output = _load_class(
```
[the remark this answers is not in the document]
Comment (nikitavemuri) [reply dashboard/modules/snapshot/snapshot_head.py:128]: We are expecting the environment variable to be set before ray is initialized, and we not supporting the use case of the environment variable itself being updated. `_load_class` loads either a class or a function from an external module and  the function is run on line 130 with `()`

File: dashboard/modules/snapshot/tests/test_snapshot.py [hunk 3]
```diff
@@ -18,12 +21,105 @@
...
+    with patch.multiple(
+        "ray.dashboard.modules.snapshot.snapshot_head", _load_class=mock_load_class
+    ):
+        response = await mock_api_head.get_component_activities(Mock(query={}))
+        data = response.body.decode()
+        expected_output = {
+            "driver": {"is_active": False, "reason": None, "timestamp": None},
+            "new_component": {"is_active": False, "reason": None, "timestamp": None},
+        }
+    os.environ.pop(RAY_CLUSTER_ACTIVITY_HOOK)
```
Comment (rkooo567) [inline dashboard/modules/snapshot/tests/test_snapshot.py:116]: Why don't we have a real e2e test here? 
```
a = {}
def endpoint():
    return {"is_active": len(a) > 0, "reason": None, "timestamp": None}

RAY_CLUSTER_ACTIVITY_HOOK=tests.test_snapshot.endpoint ray start --head

# Call API and make sure `is_active == False`

a = {"a": 1}

# call API and make sure `is_active==True`
```

**Review by rkooo567 — APPROVED** [review]
LGTM. One more test request for e2e test

**Review by rkooo567 — APPROVED** [review]

File: dashboard/modules/snapshot/snapshot_head.py [hunk 4]
```diff
@@ -108,16 +123,63 @@ async def snapshot(self, req):
...
+        # Get activity information for driver
         driver_activity_info = await self._get_job_activity_info(timeout=timeout)
-
         resp = {"driver": dataclasses.asdict(driver_activity_info)}
+
+        if RAY_CLUSTER_ACTIVITY_HOOK in os.environ:
+            try:
+                external_activity_output = _load_class(
+                    os.environ[RAY_CLUSTER_ACTIVITY_HOOK]
+                )()
```
Comment (ijrsvt) [inline dashboard/modules/snapshot/snapshot_head.py:140]:
```suggestion
                cluster_activity_callable = _load_class(
                    os.environ[RAY_CLUSTER_ACTIVITY_HOOK]
                )
                external_activity_output = cluster_activity_callable()
```
This is probably not linted correctly, but maybe making it more explicit about what is going on.

Comment (nikitavemuri) [reply dashboard/modules/snapshot/snapshot_head.py]: sure, updated

**Review by ijrsvt — APPROVED** [review]
Overall LGTM. 

The backwards incompatible change should be fine since this field was only introduced a week ago (and we haven't released Ray in the last week)

**Pull request merged**